File: Question 0417 - Pacific Atlantic Water Flow/Solution.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {

    private void dfs(int r, int c, int prev, int[][] heights, boolean[][] vis, int[][] dir) {
        if (r < 0 || r >= heights.length || c < 0 || c >= heights[0].length || vis[r][c] || prev > heights[r][c]) return;

        vis[r][c] = true;

        for (int[] d : dir) {
            int x = r + d[0];
            int y = c + d[1];

            dfs(x, y, heights[r][c], heights, vis, dir);
        }
    }

    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        List<List<Integer>> result = new ArrayList<>();

        int n = heights.length, m = heights[0].length;
        boolean[][] pac = new boolean[n][m];
        boolean[][] atl = new boolean[n][m];

        int[][] dir = { { -1, 0 }, { 1, 0 }, { 0, 1 }, { 0, -1 } };

        for (int i = 0; i < m; i++) {
            dfs(0, i, heights[0][i], heights, pac, dir);
            dfs(n - 1, i, heights[n - 1][i], heights, atl, dir);
        }

        for (int i = 0; i < n; i++) {
            dfs(i, 0, heights[i][0], heights, pac, dir);
            dfs(i, m - 1, heights[i][m - 1], heights, atl, dir);
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (pac[i][j] && atl[i][j]) result.add(Arrays.asList(i, j));
            }
        }

        return result;
    }
}

File: Question 0417 - Pacific Atlantic Water Flow/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
private:
    void dfs(vector<vector<int>> &matrix, int r, int c, int prev, vector<vector<bool>> &ocean, vector<vector<int>> &dir)
    {
        if (r < 0 || r >= matrix.size() || c < 0 || c >= matrix[0].size())
            return;

        if (ocean[r][c] || matrix[r][c] < prev)
            return;

        ocean[r][c] = true;
        for (auto &d : dir)
        {
            int x = r + d[0];
            int y = c + d[1];

            dfs(matrix, x, y, matrix[r][c], ocean, dir);
        }
    }

public:
    vector<vector<int>> pacificAtlantic(vector<vector<int>> &heights)
    {
        vector<vector<int>> ans;

        int n = heights.size(), m = heights[0].size();
        vector<vector<bool>> pac(n, vector<bool>(m, false));
        vector<vector<bool>> atlantic(n, vector<bool>(m, false));

        vector<vector<int>> dir = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        for (int i = 0; i < n; i++)
        {
            dfs(heights, i, 0, heights[i][0], pac, dir);
            dfs(heights, i, m - 1, heights[i][m - 1], atlantic, dir);
        }

        for (int i = 0; i < m; i++)
        {
            dfs(heights, 0, i, heights[0][i], pac, dir);
            dfs(heights, n - 1, i, heights[n - 1][i], atlantic, dir);
        }

        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                if (pac[i][j] && atlantic[i][j])
                    ans.push_back({i, j});
            }
        }

        return ans;
    }
};